### PR TITLE
add vendor packages filter

### PR DESCRIPTION
**Pull request by juanvallejo**
Adds graph filter

### CODE REVIEW

File: tools/depcheck/pkg/cmd/trace/filter.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+	"strings"
+
+	"github.com/gonum/graph"
+
+	depgraph "github.com/openshift/origin/tools/depcheck/pkg/graph"
+)
+
+// FilterVendorPackages receives a graph and a set of vendorRootNames contained within the graph.
+// Returns a new graph with the sub-tree for each vendorRoot collapsed into just the vendorRoot.
+// Relationships among vendorRoots are kept: if a transitive closure exists between two
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/trace/filter.go]: Pretty sure that transitive closure refers to the set of all reachable nodes, not a navigable path.

In any case, future readers will probably prefer: "edges originating from vendorRoot subpackages are rewritten to originate from the vendorRoot and edges terminating at vendorRoot subpackages are rewritten to terminate at the vendorRoot"

File: tools/depcheck/pkg/cmd/trace/filter.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+// vendor sub-trees, an edge is created between their roots in the newly collapsed graph.
+func FilterVendorPackages(g *depgraph.MutableDirectedGraph, vendorRootNames []string) (*depgraph.MutableDirectedGraph, error) {
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/trace/filter.go]: this isn't actually a vendorRootName. This is a list of "allowedNodes" or "packagePrefix" or something like that.

File: tools/depcheck/pkg/cmd/trace/filter.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+		seenClosures := map[string]bool{}
+		flatDepSet := flatten(g, r)
+
+		for _, n := range flatDepSet {
+			node, ok := n.(*depgraph.Node)
+			if !ok {
+				continue
+			}
+
+			for _, vRoot := range vRoots {
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/trace/filter.go]: Not that I won't accept n^3, but it's a little hard to read and it's hard for me to figure out if you've gotten it right.  This feels like it should be n^2 problem on a straight traversal.

File: tools/depcheck/pkg/cmd/trace/filter.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+// Relationships among vendorRoots are kept: if a transitive closure exists between two
+// vendor sub-trees, an edge is created between their roots in the newly collapsed graph.
+func FilterVendorPackages(g *depgraph.MutableDirectedGraph, vendorRootNames []string) (*depgraph.MutableDirectedGraph, error) {
+	pkgs := &PackageList{
+		Packages: []Package{},
+	}
+
+	vRoots := []*depgraph.Node{}
+
+	for _, n := range g.Nodes() {
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/trace/filter.go]: What if we built every node right here.

```go
func getFilteredNodeName(collapsedPrefixes map[string]bool, packageName string) string{
}

collapsedGraph = new

for every node in g{
    collapsedNodeName := getFilteredNodeName(collapsedPrefixes, node.package)
    add new collapsedNodeName to collapsedGraph
}

// at this point, all possible nodes are created.  Anything without a prefix got assigned an exact node match
// but we have no edges

for every node in g{
    collapsedFromNodeName := getFilteredNodeName(collapsedPrefixes, node.package)
    for all outbound edges of current node{
        collapsedToNodeName := getFilteredNodeName(collapsedPrefixes, outboundedge.to.node.package)
        add edge in collapsed graph from and to
    }
}

return the new graph.

```

File: tools/depcheck/pkg/cmd/trace/filter.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+	// copy all nodes to new graph
+	for _, n := range g.Nodes() {
+		node, ok := n.(*depgraph.Node)
+		if !ok {
+			continue
+		}
+
+		collapsedNodeName := getFilteredNodeName(packagePrefixes, node.UniqueName)
+		collapsedGraph.AddNode(&depgraph.Node{
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/trace/filter.go]: duplicates are ignored?

Comment (deads2k) [reply tools/depcheck/pkg/cmd/trace/filter.go]: > duplicates are ignored?

I expected to see a Get followed by "if not found (nil?), add"

Comment (deads2k) [reply tools/depcheck/pkg/cmd/trace/filter.go]: > duplicates are ignored?

Don't ignore.  You'll end up stuck chasing weird bugs forever.  Return an error.  You could also check for a structured error if you like.

File: tools/depcheck/pkg/cmd/trace/filter.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+			if fromNodeName == toNodeName {
+				continue
+			}
+
+			toNode, exists := collapsedGraph.NodeByName(toNodeName)
+			if !exists {
+				return nil, fmt.Errorf("expected node with name %q to exist in collapsed graph", toNodeName)
+			}
+
+			collapsedGraph.SetEdge(concrete.Edge{
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/trace/filter.go]: Same error on dupe logic here I think.

File: tools/depcheck/pkg/cmd/trace/filter_test.go [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+		"github.com/test/repo/vendor/github.com/google/glog",
+		"github.com/test/repo/vendor/github.com/docker/docker-test-util",
+	}
+
+	filteredGraph, err := FilterVendorPackages(g, vendorRoots)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// should only have vendorRoots after being filtered
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/trace/filter_test.go]: not true. Add a node not under any of your vendor roots and update your test

File: tools/depcheck/pkg/cmd/trace/filter_test.go [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+				"fmt",
+				"io",
+				"io/util",
+			},
+		},
+	},
+}
+
+func TestVendorPackagesCollapsedIntoRepo(t *testing.T) {
+	g, err := BuildGraph(vendorPkgs, []string{})
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/trace/filter_test.go]: Start this test based on the graph, not based on the original pkgs.  We want discrete unit tests.

**Comment by deads2k** [comment]
Fairly minor comments. Definitely gaining on it.

File: tools/depcheck/pkg/cmd/trace/filter.go (hunk 5 above)
Comment (juanvallejo) [reply tools/depcheck/pkg/cmd/trace/filter.go]: @deads2k 
> duplicates are ignored?

Yeah, the MutableDirectedGraph#AddNode method ignores duplicates

> Don't ignore. You'll end up stuck chasing weird bugs forever. Return an error. You could also check for a structured error if you like.

gonum's concrete.DirectedGraph panics if I try adding a node with a duplicate ID. I could do the same on MutableDirectedGraph#AddNode and panic there if I see a node with a duplicate UniqueName.

Will handle not adding duplicates to the graph in this function

File: tools/depcheck/pkg/cmd/trace/filter_test.go (hunk 7 above)
Comment (juanvallejo) [reply tools/depcheck/pkg/cmd/trace/filter_test.go]: I test non-vendor nodes in the test below

File: tools/depcheck/pkg/graph/graph.go [hunk 9]
```diff
@@ -46,6 +46,10 @@ type MutableDirectedGraph struct {
 }
 
 func (g *MutableDirectedGraph) AddNode(n *Node) {
+	if _, exists := g.nodesByName[n.UniqueName]; exists {
+		panic(fmt.Sprintf("unique: node UniqueName collision: %s", n.UniqueName))
```
Comment (deads2k) [inline tools/depcheck/pkg/graph/graph.go]: no.  return an error.

File: tools/depcheck/pkg/cmd/trace/filter.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+	for _, n := range g.Nodes() {
+		node, ok := n.(*depgraph.Node)
+		if !ok {
+			continue
+		}
+
+		collapsedNodeName := getFilteredNodeName(packagePrefixes, node.UniqueName)
+		_, exists := collapsedGraph.NodeByName(collapsedNodeName)
+		if exists {
+			continue
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/trace/filter.go]: ah, I see

**Comment by deads2k** [comment]
return an error, squash, and lgtm

**Comment by juanvallejo** [comment]
@deads2k thanks, fixed up tests, returned an error when adding a node with a duplicate name, and squashed

File: tools/depcheck/pkg/cmd/trace/filter.go [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+		}
+
+		err := collapsedGraph.AddNode(&depgraph.Node{
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/trace/filter.go]: This should have rippled to at least one other spot in the buildgraph method.

File: tools/depcheck/pkg/cmd/trace/filter_test.go [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+type testNode struct {
+	name     string
+	children []string
+}
+
+func getVendorNodes() []*testNode {
+	return []*testNode{
+		{
+			name: "github.com/test/repo/vendor/github.com/testvendor/vendor_one",
+			children: []string{
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/trace/filter_test.go]: children?  perhaps `outboundEdges`

You know, you don't get paid extra for long names.....

File: tools/depcheck/pkg/cmd/trace/filter_test.go [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+				"github.com/test/repo/vendor/github.com/testvendor/vendor_one/misc",
+			},
+		},
+		{
+			name: "github.com/test/repo/vendor/github.com/testvendor/vendor_one/misc",
+			children: []string{
+				"github.com/test/repo/vendor/github.com/testvendor/vendor_three/vendor_three_misc",
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/trace/filter_test.go]: couldn't these be like: `prefix3/three`?  Just saying, it's a lot to read to get to that final bit of entropy.

**Comment by deads2k** [comment]
/approve

minor comments and nits.  Update and you can self-apply lgtm

**Pull request merged**